components/custom-contract/tabs/code.tsx

```diff
@@ -32,41 +35,86 @@ export const CustomContractCodeTab: React.VFC<ContentOverviewProps> = ({
     )
     .map((f) => f.signature);
 
-  console.log(features);
+  useMemo(() => {
+    const generateTags = (
+      features: Record<string, FeatureWithEnabled> | undefined,
+      tags: JSX.Element[],
+    ) => {
+      if (features) {
+        Object.keys(features)
+          .map((f) => features[f])
+          .forEach((f) => {
+            tags.push(
+              <Tag
+                size="lg"
+                key={f.name}
+                borderRadius="full"
+                variant={f === expandedFeature ? "solid" : "subtle"}
+                colorScheme={f === expandedFeature ? "primary" : "gray"}
+                onClick={() => setExpandedFeature(f)}
+              >
+                <TagLabel>{f.name}</TagLabel>
+                <TagRightIcon
+                  as={f.enabled ? MdCheckCircle : IoMdCloseCircle}
+                  color={f.enabled ? "green.600" : "red.600"}
+                />
+              </Tag>,
+            );
+            // recurse to get nested features
+            generateTags(f.features, tags);
+          });
+      }
+    };
+    const tags: JSX.Element[] = [];
+    generateTags(contractFeatures, tags);
+    setFeatureTags(tags);
+  }, [contractFeatures, expandedFeature]);
```

Comment:
this is wild...

just return `tags` from the useMemo and use it (no need for state at all)